File: src/world/persistence.py

import json
import logging
import numpy as np
from pathlib import Path
from typing import Dict, Optional
from concurrent.futures import Future, ThreadPoolExecutor
from .rle import rle_encode, rle_decode

try:
    import zstandard as zstd
    has_zstd=True
except Exception:
    has_zstd=False
try:
    import lz4.frame as lz4f
    has_lz4=True
except Exception:
    has_lz4=False

class ChunkStore:
    def __init__(self, root="world_save", codec="zstd", use_rle=True, threads=4):
        self.root = Path(root)
        self.root.mkdir(parents=True, exist_ok=True)
        self.codec = codec
        self.use_rle = use_rle
        self.pool = ThreadPoolExecutor(max_workers=max(1, threads))
        self.futures: Dict[str, Future] = {}
        self.manifest_path = self.root / "manifest.json"
        if self.manifest_path.exists():
            try:
                self.manifest = json.loads(self.manifest_path.read_text())
            except Exception:
                self.manifest = {}
        else:
            self.manifest = {}

    def _region_dir(self, cx, cz):
        d = self.root / f"r.{cx//32}.{cz//32}"; d.mkdir(exist_ok=True); return d
    def chunk_path(self, cx, cz): return self._region_dir(cx, cz) / f"c.{cx}.{cz}.bin"

    def _compress(self, b: bytes) -> bytes:
        if self.codec=="zstd" and has_zstd:
            c = zstd.ZstdCompressor(level=10); return c.compress(b)
        if self.codec=="lz4" and has_lz4:
            return lz4f.compress(b, compression_level=9, block_size=lz4f.BLOCKSIZE_MAX1MB, content_checksum=True)
        return b

    def _decompress(self, b: bytes) -> bytes:
        if self.codec=="zstd" and has_zstd:
            d = zstd.ZstdDecompressor(); return d.decompress(b)
        if self.codec=="lz4" and has_lz4:
            return lz4f.decompress(b)
        return b

    def save_chunk_sync(self, chunk):
        try:
            cx, cz = chunk.position
            vox = chunk.voxels.astype(np.uint8)
            if self.use_rle:
                vals, counts, shape = rle_encode(vox)
                header = np.array([shape[0], shape[1], shape[2], vals.size, counts.size], dtype=np.int32).tobytes()
                payload = vals.tobytes() + counts.tobytes()
                data = header + payload
            else:
                header = np.array([vox.shape[0], vox.shape[1], vox.shape[2], 0, 0], dtype=np.int32).tobytes()
                data = header + vox.tobytes()
            blob = self._compress(data)
            path = self.chunk_path(cx, cz)
            path.write_bytes(blob)
            idx = self._region_dir(cx, cz) / "index.json"
            table = {}
            if idx.exists():
                try:
                    table = json.loads(idx.read_text())
                except Exception:
                    table = {}
            table[f"{cx},{cz}"] = {"saved": True}
            idx.write_text(json.dumps(table))
            region = f"r.{cx//32}.{cz//32}"
            entry = self.manifest.get(region, {"codec": self.codec, "chunks": {}})
            entry["codec"] = self.codec
            entry["chunks"][f"{cx},{cz}"] = {"offset": 0}
            self.manifest[region] = entry
            self.manifest_path.write_text(json.dumps(self.manifest))
        except Exception as exc:
            raise RuntimeError(f"failed to save chunk at {chunk.position}") from exc

    def save_async(self, chunk):
        """Schedule a chunk to be saved on a background thread.

        Callers must subsequently invoke :meth:`wait_all` and handle any
        exceptions raised during the save operations.
        """

        key = f"{chunk.position[0]},{chunk.position[1]}"
        self.futures[key] = self.pool.submit(self.save_chunk_sync, chunk)

    def load_chunk(self, cx, cz) -> Optional[Dict]:
        p = self.chunk_path(cx, cz)
        if not p.exists(): return None
        blob = p.read_bytes()
        raw = self._decompress(blob)
        header = np.frombuffer(raw[:20], dtype=np.int32)
        H, Y, S, nvals, ncnt = header.tolist()
        rest = raw[20:]
        if nvals>0:
            vals = np.frombuffer(rest[:nvals], dtype=np.uint8)
            counts = np.frombuffer(rest[nvals:nvals+4*ncnt], dtype=np.int32)
            vox = rle_decode(vals, counts, (H,Y,S))
        else:
            vox = np.frombuffer(rest, dtype=np.uint8).reshape((H,Y,S))
        return {"voxels": vox, "height": int(Y), "size": int(S)}

    def wait_all(self):
        errors = []
        for f in list(self.futures.values()):
            try:
                f.result(timeout=5)
            except Exception as exc:
                logging.exception("chunk save failed: %s", exc)
                errors.append(exc)
        self.futures.clear()
        self.pool.shutdown(wait=True)
        if errors:
            raise RuntimeError(f"{len(errors)} chunk saves failed") from errors[0]

    def delete_chunk(self, cx: int, cz: int) -> None:
        """Remove chunk data and update manifest/index."""
        region = f"r.{cx//32}.{cz//32}"
        reg_dir = self.root / region
        path = reg_dir / f"c.{cx}.{cz}.bin"
        if path.exists():
            path.unlink()
        entry = self.manifest.get(region)
        if entry and "chunks" in entry:
            entry["chunks"].pop(f"{cx},{cz}", None)
            if not entry["chunks"]:
                self.manifest.pop(region, None)
                reg_dir = self.root / region
                if reg_dir.exists():
                    for f in reg_dir.glob("*"):
                        f.unlink()
                    reg_dir.rmdir()
            self.manifest_path.write_text(json.dumps(self.manifest))
        idx = reg_dir / "index.json"
        if idx.exists():
            try:
                table = json.loads(idx.read_text())
            except Exception:
                table = {}
            table.pop(f"{cx},{cz}", None)
            if table:
                idx.write_text(json.dumps(table))
            else:
                idx.unlink()

    def compact(self) -> None:
        """Delete stray chunk files not referenced in the manifest."""
        for region, entry in list(self.manifest.items()):
            reg_dir = self.root / region
            live = set(entry.get("chunks", {}).keys())
            if reg_dir.exists():
                for chunk_file in reg_dir.glob("c.*.*.bin"):
                    coords = chunk_file.stem[2:].split(".")
                    key = f"{coords[0]},{coords[1]}"
                    if key not in live:
                        chunk_file.unlink()
            if not live and reg_dir.exists():
                for f in reg_dir.glob("*"):
                    f.unlink()
                reg_dir.rmdir()
                self.manifest.pop(region)
        self.manifest_path.write_text(json.dumps(self.manifest))

File: src/world/persistence.cpp
#include "persistence.hpp"

#include <fstream>
#include <stdexcept>
#include <iterator>
#include <algorithm>
#include <iostream>

#if __has_include(<zstd.h>)
#define VOXELVK_HAS_ZSTD 1
#include <zstd.h>
#else
#define VOXELVK_HAS_ZSTD 0
#endif

#if __has_include(<lz4frame.h>)
#define VOXELVK_HAS_LZ4 1
#include <lz4frame.h>
#else
#define VOXELVK_HAS_LZ4 0
#endif

namespace world {

ChunkStore::ChunkStore(const std::string &root, const std::string &codec, bool use_rle, int threads)
    : root_(root), codec_(codec), use_rle_(use_rle), thread_count_(threads) {
    std::filesystem::create_directories(root_);
}

ChunkStore::~ChunkStore() { wait_all(); }

std::filesystem::path ChunkStore::_region_dir(int cx, int cz) const {
    std::filesystem::path d = root_ / ("r." + std::to_string(cx / 32) + "." + std::to_string(cz / 32));
    std::filesystem::create_directories(d);
    return d;
}

std::filesystem::path ChunkStore::chunk_path(int cx, int cz) const {
    return _region_dir(cx, cz) / ("c." + std::to_string(cx) + "." + std::to_string(cz) + ".bin");
}

void ChunkStore::rle_encode(const std::vector<std::uint8_t> &arr,
                            std::vector<std::uint8_t> &vals,
                            std::vector<std::int32_t> &counts) {
    vals.clear();
    counts.clear();
    if (arr.empty()) return;
    std::uint8_t prev = arr[0];
    std::int32_t cnt = 1;
    for (std::size_t i = 1; i < arr.size(); ++i) {
        std::uint8_t v = arr[i];
        if (v == prev) {
            ++cnt;
        } else {
            vals.push_back(prev);
            counts.push_back(cnt);
            prev = v;
            cnt = 1;
        }
    }
    vals.push_back(prev);
    counts.push_back(cnt);
}

std::vector<std::uint8_t> ChunkStore::rle_decode(const std::vector<std::uint8_t> &vals,
                                                  const std::vector<std::int32_t> &counts,
                                                  std::size_t total) {
    std::vector<std::uint8_t> out;
    out.reserve(total);
    for (std::size_t i = 0; i < vals.size(); ++i) {
        out.insert(out.end(), counts[i], vals[i]);
    }
    return out;
}

std::vector<std::uint8_t> ChunkStore::_compress(const std::vector<std::uint8_t> &data) const {
    if (codec_ == "zstd" && VOXELVK_HAS_ZSTD) {
        size_t max_size = ZSTD_compressBound(data.size());
        std::vector<std::uint8_t> out(max_size);
        size_t csize = ZSTD_compress(out.data(), max_size, data.data(), data.size(), 10);
        if (ZSTD_isError(csize)) {
            throw std::runtime_error("zstd compress failed");
        }
        out.resize(csize);
        return out;
    }
#if VOXELVK_HAS_LZ4
    if (codec_ == "lz4") {
        size_t max_size = LZ4F_compressFrameBound(data.size(), nullptr);
        std::vector<std::uint8_t> out(max_size);
        size_t csize = LZ4F_compressFrame(out.data(), max_size, data.data(), data.size(), nullptr);
        if (LZ4F_isError(csize)) {
            throw std::runtime_error("lz4 compress failed");
        }
        out.resize(csize);
        return out;
    }
#endif
    return data;
}

std::vector<std::uint8_t> ChunkStore::_decompress(const std::vector<std::uint8_t> &data) const {
    if (codec_ == "zstd" && VOXELVK_HAS_ZSTD) {
        unsigned long long size = ZSTD_getFrameContentSize(data.data(), data.size());
        if (size == ZSTD_CONTENTSIZE_ERROR || size == ZSTD_CONTENTSIZE_UNKNOWN) {
            throw std::runtime_error("zstd frame size unknown");
        }
        std::vector<std::uint8_t> out(size);
        size_t dsize = ZSTD_decompress(out.data(), size, data.data(), data.size());
        if (ZSTD_isError(dsize)) {
            throw std::runtime_error("zstd decompress failed");
        }
        return out;
    }
#if VOXELVK_HAS_LZ4
    if (codec_ == "lz4") {
        LZ4F_dctx *dctx;
        LZ4F_createDecompressionContext(&dctx, LZ4F_VERSION);
        LZ4F_frameInfo_t info{};
        size_t consumed = 0;
        size_t src_size = data.size();
        const uint8_t *src = data.data();
        LZ4F_getFrameInfo(dctx, &info, src, &consumed);
        src += consumed;
        src_size -= consumed;
        if (info.contentSize == 0) {
            LZ4F_freeDecompressionContext(dctx);
            throw std::runtime_error("lz4 unknown content size");
        }
        std::vector<std::uint8_t> out(info.contentSize);
        size_t dst_size = out.size();
        size_t result = LZ4F_decompress(dctx, out.data(), &dst_size, src, &src_size, nullptr);
        LZ4F_freeDecompressionContext(dctx);
        if (LZ4F_isError(result)) {
            throw std::runtime_error("lz4 decompress failed");
        }
        return out;
    }
#endif
    return data;
}

void ChunkStore::save_chunk_sync(const Chunk &chunk) {
    int cx = chunk.position.first;
    int cz = chunk.position.second;
    const auto &vox = chunk.voxels;
    if (static_cast<std::size_t>(chunk.size * chunk.height * chunk.size) != vox.size()) {
        throw std::runtime_error("voxel array size mismatch");
    }
    std::vector<std::uint8_t> vals;
    std::vector<std::int32_t> counts;
    std::vector<std::uint8_t> data;

    std::int32_t header[5] = {chunk.height, chunk.size, chunk.size, 0, 0};
    if (use_rle_) {
        rle_encode(vox, vals, counts);
        header[3] = static_cast<std::int32_t>(vals.size());
        header[4] = static_cast<std::int32_t>(counts.size());
    }

    data.insert(data.end(), reinterpret_cast<std::uint8_t *>(header),
                reinterpret_cast<std::uint8_t *>(header) + sizeof(header));
    if (use_rle_) {
        data.insert(data.end(), vals.begin(), vals.end());
        data.insert(data.end(), reinterpret_cast<const std::uint8_t *>(counts.data()),
                    reinterpret_cast<const std::uint8_t *>(counts.data()) + counts.size() * sizeof(std::int32_t));
    } else {
        data.insert(data.end(), vox.begin(), vox.end());
    }
    auto blob = _compress(data);
    auto path = chunk_path(cx, cz);
    std::ofstream ofs(path, std::ios::binary);
    ofs.write(reinterpret_cast<const char *>(blob.data()), static_cast<std::streamsize>(blob.size()));
    ofs.close();

    // Minimal index.json write
    auto idx = _region_dir(cx, cz) / "index.json";
    std::ofstream idxf(idx, std::ios::binary);
    idxf << "{\"" << cx << "," << cz << "\": {\"saved\": true}}";
}

void ChunkStore::save_async(const Chunk &chunk) {
    futures_.emplace_back(std::async(std::launch::async, [this, chunk] { save_chunk_sync(chunk); }));
}

std::optional<ChunkData> ChunkStore::load_chunk(int cx, int cz) {
    auto path = chunk_path(cx, cz);
    if (!std::filesystem::exists(path)) return std::nullopt;
    std::ifstream ifs(path, std::ios::binary);
    std::vector<std::uint8_t> blob((std::istreambuf_iterator<char>(ifs)), std::istreambuf_iterator<char>());
    ifs.close();
    auto raw = _decompress(blob);
    if (raw.size() < sizeof(std::int32_t) * 5) return std::nullopt;
    const std::int32_t *header = reinterpret_cast<const std::int32_t *>(raw.data());
    int height = header[0];
    int size_x = header[1];
    int size_z = header[2];
    int nvals = header[3];
    int ncnt = header[4];
    std::size_t total = static_cast<std::size_t>(height) * size_x * size_z;
    const std::uint8_t *ptr = raw.data() + sizeof(std::int32_t) * 5;
    std::vector<std::uint8_t> vox;
    if (nvals > 0) {
        if (raw.size() < sizeof(std::int32_t) * 5 + static_cast<std::size_t>(nvals) + static_cast<std::size_t>(ncnt) * sizeof(std::int32_t)) {
            return std::nullopt;
        }
        std::vector<std::uint8_t> vals(ptr, ptr + nvals);
        const std::int32_t *cnt_ptr = reinterpret_cast<const std::int32_t *>(ptr + nvals);
        std::vector<std::int32_t> counts(cnt_ptr, cnt_ptr + ncnt);
        vox = rle_decode(vals, counts, total);
    } else {
        if (raw.size() < sizeof(std::int32_t) * 5 + total) return std::nullopt;
        vox.assign(ptr, ptr + total);
    }
    return ChunkData{std::move(vox), height, size_x};
}

void ChunkStore::wait_all() {
    for (auto &f : futures_) {
        try {
            f.get();
        } catch (const std::exception &e) {
            std::cerr << "[ChunkStore::wait_all] Exception in async save: " << e.what() << std::endl;
        } catch (...) {
            std::cerr << "[ChunkStore::wait_all] Unknown exception in async save." << std::endl;
        }
    }
    futures_.clear();
}

} // namespace world

